File: src/simulator/paper_broker.py
from dataclasses import dataclass
from datetime import datetime, timezone


@dataclass
class Trade:
    timestamp_utc: str
    symbol: str
    side: str
    quantity: int
    price: float
    notional: float
    fee: float
    realized_pnl: float
    reason: str


class PaperBroker:
    def __init__(self, initial_capital: float = 10000.0, fee_rate: float = 0.0) -> None:
        self.cash = initial_capital
        self.fee_rate = fee_rate
        self.positions: dict[str, int] = {}
        self.avg_cost: dict[str, float] = {}
        self.realized_pnl = 0.0

    def execute(self, symbol: str, side: str, quantity: int, price: float, reason: str) -> Trade | None:
        if quantity <= 0 or price <= 0:
            return None

        ts = datetime.now(timezone.utc).isoformat()
        notional = quantity * price
        fee = notional * self.fee_rate
        realized = 0.0

        if side == "BUY":
            total_cost = notional + fee
            if total_cost > self.cash:
                return None
            prev_qty = self.positions.get(symbol, 0)
            prev_cost = self.avg_cost.get(symbol, 0.0)
            new_qty = prev_qty + quantity
            new_avg = ((prev_qty * prev_cost) + notional) / new_qty
            self.positions[symbol] = new_qty
            self.avg_cost[symbol] = new_avg
            self.cash -= total_cost

        elif side == "SELL":
            prev_qty = self.positions.get(symbol, 0)
            if quantity > prev_qty:
                return None
            avg = self.avg_cost.get(symbol, 0.0)
            realized = (price - avg) * quantity - fee
            self.realized_pnl += realized
            self.positions[symbol] = prev_qty - quantity
            if self.positions[symbol] == 0:
                self.avg_cost.pop(symbol, None)
            self.cash += notional - fee
        else:
            return None

        return Trade(
            timestamp_utc=ts,
            symbol=symbol,
            side=side,
            quantity=quantity,
            price=price,
            notional=notional,
            fee=fee,
            realized_pnl=realized,
            reason=reason,
        )

File: src_cpp/simulator/paper_broker.cpp
#include "simulator/paper_broker.hpp"

#include <chrono>
#include <ctime>
#include <iomanip>
#include <sstream>

PaperBroker::PaperBroker(double initial_capital, double fee_rate, double max_position_notional)
    : cash_(initial_capital), fee_rate_(fee_rate), max_position_notional_(max_position_notional) {}

std::string PaperBroker::now_utc() const {
    const auto now = std::chrono::system_clock::now();
    const std::time_t t = std::chrono::system_clock::to_time_t(now);
    std::tm tm_utc {};
#ifdef _WIN32
    gmtime_s(&tm_utc, &t);
#else
    gmtime_r(&t, &tm_utc);
#endif
    std::ostringstream oss;
    oss << std::put_time(&tm_utc, "%Y-%m-%dT%H:%M:%SZ");
    return oss.str();
}

Trade PaperBroker::execute(const std::string& symbol, const std::string& side, int quantity, double price, const std::string& reason, bool& ok) {
    ok = false;
    Trade t;
    t.timestamp_utc = now_utc();
    t.symbol = symbol;
    t.side = side;
    t.quantity = quantity;
    t.price = price;
    t.notional = quantity * price;
    t.fee = t.notional * fee_rate_;
    t.reason = reason;

    if (quantity <= 0 || price <= 0.0) {
        t.reason = "Rejected: invalid quantity or price";
        return t;
    }

    if (side == "BUY") {
        const double current_notional = static_cast<double>(positions_[symbol]) * price;
        if (current_notional + t.notional > max_position_notional_) {
            t.reason = "Rejected: max position notional exceeded";
            return t;
        }

        const double total_cost = t.notional + t.fee;
        if (total_cost > cash_) {
            t.reason = "Rejected: insufficient cash";
            return t;
        }

        const int prev_qty = positions_[symbol];
        const double prev_avg = avg_cost_[symbol];
        const int new_qty = prev_qty + quantity;
        const double new_avg = ((prev_qty * prev_avg) + t.notional) / static_cast<double>(new_qty);
        positions_[symbol] = new_qty;
        avg_cost_[symbol] = new_avg;
        cash_ -= total_cost;
        ok = true;
        return t;
    }

    if (side == "SELL") {
        const int prev_qty = positions_[symbol];
        if (quantity > prev_qty) {
            t.reason = "Rejected: insufficient position";
            return t;
        }

        const double avg = avg_cost_[symbol];
        t.realized_pnl = (price - avg) * quantity - t.fee;
        realized_pnl_ += t.realized_pnl;
        positions_[symbol] = prev_qty - quantity;
        if (positions_[symbol] == 0) {
            avg_cost_.erase(symbol);
        }
        cash_ += t.notional - t.fee;
        ok = true;
        return t;
    }

    t.reason = "Rejected: unsupported side";
    return t;
}
